patchwork-registries/src/main/java/net/minecraftforge/registries/RegistryManager.java

```diff
@@ -43,28 +53,121 @@ public String getName() {
 
 	@SuppressWarnings("unchecked")
 	public <V extends IForgeRegistryEntry<V>> Class<V> getSuperType(Identifier key) {
-		return (Class<V>) RegistryClassMapping.getClass(key);
+		return (Class<V>) this.superTypes.inverse().get(key);
 	}
 
 	@SuppressWarnings("unchecked")
 	public <V extends IForgeRegistryEntry<V>> ForgeRegistry<V> getRegistry(Identifier key) {
-		return (ForgeRegistry<V>) this.registries.get(key);
+		Registry<?> vanillaRegistry = Registry.REGISTRIES.get(key);
+
+		if (vanillaRegistry instanceof ForgeRegistryProvider) {
+			return ((ForgeRegistryProvider) vanillaRegistry).getForgeRegistry();
+		}
+
+		return null;
 	}
 
 	public <V extends IForgeRegistryEntry<V>> IForgeRegistry<V> getRegistry(Class<? super V> clazz) {
-		return getRegistry(RegistryClassMapping.getIdentifier(clazz));
+		Identifier existingKey = this.getName(clazz);
+
+		if (existingKey == null) {
+			return null;
+		}
+
+		return getRegistry(existingKey);
 	}
 
 	public <V extends IForgeRegistryEntry<V>> Identifier getName(IForgeRegistry<V> reg) {
 		return reg.getRegistryName();
 	}
 
-	/**
-	 * Used by {@link net.minecraftforge.registries.ForgeRegistries}.
-	 *
-	 * @param registry the registry to add to the mapping
-	 */
-	void addRegistry(Identifier key, ForgeRegistry registry) {
-		registries.put(key, registry);
+	public Identifier getName(Class<?> clazz) {
+		if (clazz == null) {
+			return null;
+		}
+
+		do {
```

Comment:
Ew.
Would changing this to a normal while loop and removing the extra null check above break any behavior in this method?

Reply:
Object.class.getSuperclass() returns null, so that part stays the same. But, the existingKey code would be missed. Is there any reason that the existingKey stuff isn't the first thing in the method?

Reply:
Okay, after looking closely, it would only change the behavior if there's a registry with a super type of Object. That shouldn't happen, and in fact, we should probably have checks for that in the creation methods.

Reply:
I have changed the loop to a standard while loop